src/Analysis/Engine/Impl/DependencyResolution/PathResolverSnapshot.cs

```diff
@@ -0,0 +1,280 @@
+                    break;
+                }
+                lastArc = lastArc.Append(childIndex);
+            }
+
+            unmatchedPathSpan = nameSpan.start != -1 ? (nameSpan.start, modulePathLength - nameSpan.start) : (-1, 0);
+            return lastArc;
+        }
+
+        private static bool TryAppendPackages(in Arc arc, in IEnumerable<string> packageNames, out Arc lastArc) {
+            lastArc = arc;
+            foreach (var name in packageNames) {
+                if (lastArc.End.IsModule) {
+                    return false;
+                }
+                var index = lastArc.End.GetChildIndex(name);
+                if (index == -1) {
+                    return false;
+                }
+                lastArc = lastArc.Append(index);
+            }
+
+            return true;
+        }
+
+        private bool TryMatchPath(in Arc lastArc, in int rootIndex, out Arc matchedLastArc) {
+            var sourceArc = lastArc.GetFirstArc();
+            matchedLastArc = new Arc(rootIndex, _roots[rootIndex]);
+
+            while (sourceArc != lastArc) {
+                sourceArc = sourceArc.Next;
+                var childIndex = matchedLastArc.End.GetChildIndex(sourceArc.End.Name);
+                if (childIndex == -1) {
+                    return false;
+                }
+                matchedLastArc = matchedLastArc.Append(childIndex);
+            }
+
+            return true;
+        }
+
+        private static Node CreateNewNodes(string modulePath, (int start, int length) unmatchedPathSpan, Node packageNode) {
+            if (packageNode != default) {
+                // Module name matches name of existing package        
+                return packageNode.ToModule(modulePath);
+            }
+
+            if (unmatchedPathSpan.start == modulePath.LastIndexOf(Path.DirectorySeparatorChar) + 1) {
+                // Module is added to existing package
+                return Node.CreateModule(modulePath.Substring(unmatchedPathSpan.start, unmatchedPathSpan.length), modulePath);
+            }
+
+            var names = modulePath.Split(Path.DirectorySeparatorChar, unmatchedPathSpan.start, unmatchedPathSpan.length);
+            var newNode = Node.CreateModule(names.Last(), modulePath);
+
+            for (var i = names.Length - 2; i >= 0; i--) {
+                newNode = new Node(names[i], newNode);
+            }
+
+            return newNode;
+        }
+
+        private static Node UpdateNodesFromEnd(Arc lastArc, Node newEnd) {
+            while (lastArc.Start != default) {
+                var newStart = lastArc.Start.ReplaceChildAt(newEnd, lastArc.EndIndex);
+                lastArc = lastArc.Previous;
+                newEnd = newStart;
+            }
+
+            return newEnd;
+        }
+
+        private bool TryFindModule(string modulePath, out Arc lastArc, out string normalizedPath, out (int start, int length) unmatchedPathSpan) {
+            if (!Path.IsPathRooted(modulePath)) {
```

Comment:
Is it time to introduce `IFileSystem` for testability along with service container?... So we can run Mac path tests on Windows and the other way around?

Reply:
Yes it is. Do you want me to add it to this [WIP]?

Reply:
Sure, prob will need for tests anyway.